### PR TITLE
Code refactor and new translation mechanism

### PR DESCRIPTION
**Pull request by asimakiskydros**
# Description

* Code structure refactored to be more readable and contain more comments.
* App function and look should be about the same, changed how the app goes from configuration page/settings page back to main to be smoother.
* Tested for bugs, I fixed as many old ones and as many new ones as I could find.
* I never got the original translation mechanism working, I tried both the .exe and compiling the source code. I made a new one from scratch, I hope that's ok.
* Following the above, I also included Greek in the available languages. 
* Any undefined behaviors in the app's abtract logic I left open, for you to decide in the future what the program should do.

Notify me for any bugs or deviations from the app's expected function my changes may have caused.
I'm including 'Breaking change' and 'Documentation update required' below because I guess they're true, considering the
translation update.

Fixes # Code readability and maintenance, translation

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

I tested manually, as I went, testing everything again with every new change to ensure compatibility. I didn't implement any
automated tests because I didn't find it necessary (the app doesn't have many different functionalities as it stands, anyways).

**Test Configuration**:
* Firmware version: N/A
* Hardware: N/A
* Toolchain: N/A
* SDK: N/A

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by Kubis10** [comment]
@asimakiskydros 
Firstly, thanks a lot for this pr, you did a tremendous job for which I am grateful. However, I have a couple of comments:
- "My profile" window takes strangely too long to load,
- Something is wrong with the counting of messages, because in the detailed view the number of messages from specific people does not add up well and is much less than it should be,
- at line 542 this break doesn't work for me because it won't even turn on the message loading window when it's there, and it only works if I replace it with continue.
- I have error on some groups chat that title only contain numbers like "22:22" or "12563":
```
Exception in Tkinter callback
Traceback (most recent call last):
  File "tkinter\__init__.py", line 1948, in __call__
    return self.func(*args)
           ^^^^^^^^^^^^^^^^
  File "CounterForMessenger\new_main.py", line 117, in <lambda>
    self.treeview.bind('<Double-1>', lambda event: self.show_statistics())
                                                   ^^^^^^^^^^^^^^^^^^^^^^
  File "CounterForMessenger\new_main.py", line 225, in show_statistics
    StatisticsPopup(self.controller, selection[5])
  File "CounterForMessenger\new_main.py", line 577, in __init__
    # display popup title
                          
  File "CounterForMessenger\new_main.py", line 362, in extract_data
    for file in glob.glob(self.directory + conversation + '/*.json'):
                          ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
TypeError: can only concatenate str (not "int") to str
```
Other than that, I haven't seen any other problems yet

File: new_main.py [hunk 1]
```diff
@@ -0,0 +1,611 @@
...
+        self.directory = self.controller.get_directory()
+        if self.directory != '' and not self.directory.isspace() and self.directory != self.module.TITLE_NO_SELECTION:
+            for conversation in listdir(self.directory):
+                try:
+                    title, people, room, all_msgs, calltime, _, _ = self.controller.extract_data(conversation)
+                    if title == '':
+                        # if this occurs, the given path is of correct directory format but contains no useful info
+                        # (meaning it's not the expected inbox folder)
+                        # skip the entire process, nothing to show
+                        break
```
Comment (Kubis10) [inline new_main.py:541]: Because of this break, my messages are not loading at all

**Review by Kubis10 — CHANGES_REQUESTED** [review]

**Comment by asimakiskydros** [comment]
Thanks for letting me know, I'm working on these issues as we speak. 

Regarding the third one, do you mean nothing loads when you press 'Upload', having provided the correct directory path? Or when you just gave a wrong one? I'm asking this because it was my intention for nothing to be shown if the path is unexpected, it would just project garbage. If you gave a correct path and this happened, I will have to rework this entire mechanism, simply replacing break with continue will lead to undefined behavior in the future.

**Comment by asimakiskydros** [comment]
Do you happen to have chats that are titled ''? This may be a culprit for the above.

**Comment by asimakiskydros** [comment]
Ok, I made some changes, let's get the easy stuff out of the way:
* The first problem was 100% my bad, I made a rookie mistake. I reverted the counting logic back to what you had implemented.
* Following the above, the second problem should also be fixed, although I never actually encountered it in my testing. Let me know if it's still there.
* I was dumbfounded about the fourth one, turns out the interpreter will automatically convert strings to integers in that instance if possible. I changed the string format to f-string, I'd like to believe it is fixed now (works fine on my end).

About the third one, I'm gonna assume you test edge cases and you have chats that are titled blank or something. I changed the break condition to the number of participants being 0. This should of course be impossible in any context. As I said, nothing showing while a wrong path is provided is the expected behavior. Let me know what happens on your end.

**Comment by Kubis10** [comment]
I have correct directory path because only when I changed this break to continue it works perfectly. Referring to blank chats title - I have some very old group chats that are empty in name and maybe because of this, as you suggested, loading is not working.

**Comment by asimakiskydros** [comment]
> I have correct directory path because only when I changed this break to continue it works perfectly. Referring to blank chats title - I have some very old group chats that are empty in name and maybe because of this, as you suggested, loading is not working.

You keep having this problem with the new changes?

**Comment by Kubis10** [comment]
For now, I'm away from home until Friday and I don't have access to data, so I'll check thoroughly then

**Comment by Kubis10** [comment]
- My profile is now working good
- As you see below: 
image
Still this is not working
- Loading is now working for me
- Now there is no error but chat details is empty:
image

**Comment by asimakiskydros** [comment]
Ok, I'm gonna need context for both errors, as I haven't encountered either of them:
* Is there anything specific, reoccuring or out of the ordinary in the details of the chats where you encounter the first problem? For example, are all of them private chats, do they have weird titles, etc?
* The second one occured from you clicking on an existing listing right? You didn't click on empty whitespace and the window just happened to appear?

**Comment by Kubis10** [comment]
Counting doesn't work in chats that are 8k+ long for me properly, so that's strange.
Second problem appears when I clicked group chat that in normal view is correct:
image

**Comment by asimakiskydros** [comment]
I'm sorry, what does "8k+ long" mean? Also, I noticed that the group chat's name is the one with the numbers, does the same problem occur with other chats named with only numbers?

**Comment by Kubis10** [comment]
The problem with accurately counting messages for each person separately occurs in chats that have over 8.5 thousand messages, because unfortunately I don't have any conversation that would have 9 thousand, but at 10 it already occurs.

Also, yes, it appears in only number chats and I think I know why:
image
This is what folder name is and this is what app save as it name and want to read:
image

**Comment by Kubis10** [comment]
I think I know where 1 problem comes from:
Chats that have 10,000 messages are divided into files of 10,000 max and it turns out that the application only counts the last file where there is a surplus that did not fit into the previous file

**Comment by asimakiskydros** [comment]
Thanks, this is all very useful. 
The counting problem doesn't occur with the original code? I'm asking because I'm not very familiar with .JSON files and I emulated your own approach as best as I could. Is the file division problem addressed there?

**Comment by asimakiskydros** [comment]
OK, so, the ttk treeview will go out of its way to convert able strings to integers. Because Python allows udnerscores in integers, the conversion can happen. This conversion is not preventable, so the underscore must be changed to random letters when inserted to and changed back when pulled from the treeview. I'm sure you can understand that this is not an 100% perfect solution but it's the best I could come up with. Let me know if it works for you, I tried it with some manually created folders and it works.

I'm still working on the first problem, I'll keep you posted.

**Comment by Kubis10** [comment]
For me now, chats with numbers are working fine

**Comment by asimakiskydros** [comment]
Alright, so extract_data() would recreate the participants dictionary with each new .json file, discarding previous collections. This should fix the first problem.

**Comment by Kubis10** [comment]
Now everything works fine. Thanks!
I found one more bug with chats that have names either empty or consist of some strange characters like:
image
And when i want to view this I have this error:
```
Exception in Tkinter callback
Traceback (most recent call last):
  File "Python\Python311\Lib\tkinter\__init__.py", line 1948, in __call__
    return self.func(*args)
           ^^^^^^^^^^^^^^^^
  File "CounterForMessengerRe\new_main.py", line 120, in <lambda>
    self.treeview.bind('<Double-1>', lambda event: self.show_statistics())
                                                   ^^^^^^^^^^^^^^^^^^^^^^
  File "CounterForMessengerRe\new_main.py", line 224, in show_statistics
    StatisticsPopup(self.controller, selection[5].replace(DELIMITER, '_'))
                                     ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'int' object has no attribute 'replace'
```
But apart from this error, everything seems to be ready to go.

**Review by Kubis10 — APPROVED** [review]

**Comment by asimakiskydros** [comment]
I would imagine the same underscore problem is at play here. Can you please give me a screenshot of the folder names of 2 examples of these culprits, one with chat title empty and the other with these strange characters?

**Comment by Kubis10** [comment]
Empty and special characters are just numbers: 

image

**Comment by Kubis10** [comment]
Now everything works perfectly, merging changes.

**Comment by asimakiskydros** [comment]
Great! I'm glad I could help.
